packages/lesswrong/server/fmCrosspost.ts

```diff
@@ -0,0 +1,381 @@
+  route: ApiRoute,
+  body: T,
+  expectedStatus: string,
+  onErrorMessage: string,
+) => {
+  const result = await fetch(makeApiUrl(route), {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify(body),
+  });
+  const json = await result.json();
+  if (json.status !== expectedStatus) {
+    throw new ApiError(500, onErrorMessage);
+  }
+  return json;
+}
+
+const signToken = <T extends {}>(payload: T): string =>
+  jwt.sign(payload, getSecret(), jwtSigningOptions);
+
+const verifyToken = <T extends {}>(token: string): T =>
+  jwt.verify(token, getSecret()) as T;
+
+const crosspostResolvers = {
+  Mutation: {
+    connectCrossposter: async (
+      root: void,
+      {token}: ConnectCrossposterArgs,
+      {req, res}: ResolverContext,
+    ) => {
+      const localUserId = getUserId(req);
+      const {foreignUserId} = await makeCrossSiteRequest(
+        apiRoutes.connectCrossposter,
+        {token, localUserId},
+        "connected",
+        "Failed to connect accounts for crossposting",
+      );
+      await Users.rawUpdateOne({_id: localUserId}, {
+        $set: {fmCrosspostUserId: foreignUserId},
+      });
+      return "success";
+    },
+    unlinkCrossposter: async (root: void, args: {}, {req, res}: ResolverContext) => {
+      const localUserId = getUserId(req);
+      const foreignUserId = req?.user?.fmCrosspostUserId;
+      if (foreignUserId) {
+        const token = signToken<UnlinkCrossposterPayload>({userId: foreignUserId});
+        await makeCrossSiteRequest(
+          apiRoutes.unlinkCrossposter,
+          {token},
+          "unlinked",
+          "Failed to unlink crossposting accounts",
+        );
+        await Users.rawUpdateOne({_id: localUserId}, {
+          $unset: {fmCrosspostUserId: ""},
+        });
+      }
+      return "success";
+    },
+  },
+};
+
+addGraphQLResolvers(crosspostResolvers);
+addGraphQLMutation("connectCrossposter(token: String): String");
+addGraphQLMutation("unlinkCrossposter: String");
+
+const updateCrosspost = async (post: DbPost) => {
+  if (post.fmCrosspost?.foreignPostId) {
+    const token = signToken<UpdateCrosspostPayload>({
+      postId: post.fmCrosspost.foreignPostId,
+      draft: post.draft,
+      deletedDraft: post.deletedDraft,
+      title: post.title,
+    });
+    await makeCrossSiteRequest(
+      apiRoutes.updateCrosspost,
+      {token},
+      "updated",
+      "Failed to update crosspost draft status",
+    );
+  }
+}
+
+const withApiErrorHandlers = (callback: (req: Request, res: Response) => Promise<void>) =>
+  async (req: Request, res: Response) => {
+    try {
+      await callback(req, res);
+    } catch (e) {
+      res
+        .status(e instanceof ApiError ? e.code : 500)
+        .send({error: e.message ?? "An unknown error occurred"})
+    }
+  }
+
+const getPostParams = (req: Request, paramNames: string[]) => {
+  const params = paramNames.map((name) => req.body[name])
```

Comment:
It feels like we're doing a lot of fairly ad-hoc response validation in a few different ways (or not doing it in some places, and doing it in others).  This mechanism doesn't give us any type-safety - if we don't want to pull in a 3rd-party library like io-ts, can we instead just write a tiny validator function for each of the 4 routes' expected responses?